File: test_sim.py
class SimTest:
    def __init__(self):
        self.memory = bytearray(600 * 65536)
        self.flatNodes = [
            {'id': 'node1', 'type': 'RAM', 'addressPins': 8, 'memoryData': list(range(256))},
            {'id': 'node2', 'type': 'RAM', 'addressPins': 4, 'memoryData': [i + 100 for i in range(16)]}
        ]
        
    def buildExecutionGraph(self):
        currentRomOffset = 0
        for n in self.flatNodes:
            if n['type'] == 'RAM':
                allocSize = 1 << n['addressPins']
                for i in range(allocSize):
                    self.memory[16777216 + currentRomOffset + i] = n['memoryData'][i]
                currentRomOffset += allocSize

    def updateHostMemory(self):
        currentRomOffset = 0
        for fn in self.flatNodes:
            if fn['type'] == 'RAM' or fn['type'] == 'ROM':
                allocSize = 1 << fn['addressPins']
                if fn['type'] == 'RAM':
                    hostNode = next(n for n in self.flatNodes if n['id'] == fn['id'])
                    if 'memoryData' not in hostNode or len(hostNode['memoryData']) != allocSize:
                        hostNode['memoryData'] = [0] * allocSize
                    for i in range(allocSize):
                        hostNode['memoryData'][i] = self.memory[16777216 + currentRomOffset + i]
                currentRomOffset += allocSize

sim = SimTest()
sim.buildExecutionGraph()
sim.updateHostMemory()
print("Chip 1:", sim.flatNodes[0]['memoryData'][:20])
print("Chip 2:", sim.flatNodes[1]['memoryData'])

# Now simulate ADDING a second RAM chip AFTER the first one was loaded.
# The `flattenNodes` process reverses the nodes.
sim.flatNodes = [
    {'id': 'node2', 'type': 'RAM', 'addressPins': 4, 'memoryData': [i + 100 for i in range(16)]},
    sim.flatNodes[0]
]

# When second chip is added, buildExecutionGraph runs.
sim.buildExecutionGraph()
sim.updateHostMemory()
print("After adding Chip 2, Chip 1:", sim.flatNodes[1]['memoryData'][:20])
print("After adding Chip 2, Chip 2:", sim.flatNodes[0]['memoryData'])

File: test_sim.js
class SimTest {
    constructor() {
        this.memory = new WebAssembly.Memory({ initial: 600 });
        this.flatNodes = [
            { id: 'node1', type: 'RAM', addressPins: 8, memoryData: new Array(256).fill(0).map((_, i) => i) },
            { id: 'node2', type: 'RAM', addressPins: 4, memoryData: new Array(16).fill(0).map((_, i) => i + 100) }
        ];
    }
    
    buildExecutionGraph() {
        let currentRomOffset = 0;
        this.flatNodes.forEach(n => {
            if (n.type === 'RAM') {
                const allocSize = 1 << n.addressPins;
                const view = new Uint8Array(this.memory.buffer, 16777216 + currentRomOffset, allocSize);
                view.set(n.memoryData.slice(0, allocSize));
                currentRomOffset += allocSize;
            }
        });
    }

    updateHostMemory() {
        let currentRomOffset = 0;
        this.flatNodes.forEach(fn => {
            if (fn.type === 'RAM' || fn.type === 'ROM') {
                const allocSize = 1 << fn.addressPins;
                if (fn.type === 'RAM') {
                    const hostNode = this.flatNodes.find(n => n.id === fn.id);
                    if (hostNode) {
                        const view = new Uint8Array(this.memory.buffer, 16777216 + currentRomOffset, allocSize);
                        if (!hostNode.memoryData || hostNode.memoryData.length !== allocSize) hostNode.memoryData = new Array(allocSize).fill(0);
                        for(let i = 0; i < allocSize; i++) hostNode.memoryData[i] = view[i];
                    }
                }
                currentRomOffset += allocSize;
            }
        });
    }
}

const sim = new SimTest();
sim.buildExecutionGraph();
sim.updateHostMemory();
console.log("Chip 1:", sim.flatNodes[0].memoryData.slice(0, 20).join(', '));
console.log("Chip 2:", sim.flatNodes[1].memoryData.join(', '));
